File: graphs/cppRealizations/numberofConnectedComponentsInAnUndirectedGraph.cpp
#include <vector>
#include <map>
#include <set>
#include <iostream>

using namespace std;


class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        return 2;
    }
};


template <typename T>
ostream& operator<<(ostream& out, const vector<T>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}


template<typename T>
ostream& operator<<(ostream& out, const vector<vector<T>>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}


int main(){
    int n = 6;
    vector<vector<int>> edges = {{0,1},{1,2},{2,3},{4,5}};
    cout << "Input: n = " << n << ", edges = " << edges << endl;
    cout << "Solution: " << Solution().countComponents(n, edges) << endl;
    return 0;
}

File: graphs/numberofConnectedComponentsInAnUndirectedGraph.py
class UnionFind:
    def __init__(self):
        self.f = {}

    def findParent(self, x):
        y = self.f.get(x, x)
        if x != y:
            y = self.f[x] = self.findParent(y)
        return y

    def union(self, x, y):
        self.f[self.findParent(x)] = self.findParent(y)


class Solution:
    def countComponents(self, n: int, edges: list[list[int]]) -> int:
        dsu = UnionFind()
        for a, b in edges:
            dsu.union(a, b)
        return len(set(dsu.findParent(x) for x in range(n)))
    

if __name__ == "__main__":
    n = 5
    edges = [[0,1], [1,2], [2,3], [4,5]]
    print('Input: n = {}, edges = {}'.format(n, edges))
    print('Solution: {}'.format(Solution().countComponents(n, edges)))